File: go/heap/347_topk_freq.go
package heap

import "container/heap"

// topKFrequent 给你一个整数数组 nums 和一个整数 k ，请你返回其中出现频率前 k 高的元素。
//  用小根堆 解决TopK 大的问题。逻辑是长度为k的堆中，当容量满了之后，需要删除堆顶的最小的元素
// 用hashMap{key: cnt} 表示一个堆的元素
func topKFrequent(nums []int, k int) []int {

	hmap := make(map[int]int)

	for _, v := range nums {
		hmap[v] += 1
	}

	var kv = make(KVList, 0)
	heap.Init(&kv)

	for key, v := range hmap {
		heap.Push(&kv, [2]int{key, v})
		if kv.Len() > k {
			heap.Pop(&kv)
		}
	}

	ans := make([]int, 0)

	for i := 0; i < k; i++ {
		item := heap.Pop(&kv)
		ans = append(ans, item.([2]int)[0])
	}
	return ans
}

// type KV struct {
// 	k int
// 	v int
// }

type KVList [][2]int // []{[k, cnt], [k, cnt]...}

func (kv KVList) Len() int {
	return len(kv)
}

func (kv KVList) Less(i, j int) bool {
	return kv[i][1] < kv[j][1]
}

func (kv KVList) Swap(i, j int) {
	kv[i], kv[j] = kv[j], kv[i]
}

func (kv *KVList) Push(x interface{}) {
	(*kv) = append((*kv), x.([2]int))
}

func (kv *KVList) Pop() interface{} {
	item := (*kv)[len(*kv)-1]
	(*kv) = (*kv)[:len(*kv)-1]

	return item
}

File: rust/src/heap/347_topk_freq.rs

pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
    use std::collections::{BinaryHeap, HashMap};
    use std::cmp::Reverse;

    let mut hmap = HashMap::new();

    let mut mini_heap = BinaryHeap::with_capacity(k as usize);

    // 用hashMap 统计次数
    nums.iter().for_each(|&x|{
        let cnt = hmap.entry(x).or_insert(0);
        *cnt += 1;
    });

    // 将{num: 次数} 用小堆根存储起来
    for (key, val) in hmap {

        if mini_heap.len() == mini_heap.capacity() {
            // Reverse 反转比较器 Reverse(i) > Reverse(j) means i < j
            // 查看次数是否需要加入堆中, 小根堆topk 频繁，堆顶小于freq val 应该加入堆中
            // reverse 反转这个情况
            if *mini_heap.peek().unwrap() > Reverse((val, key)) {
                mini_heap.pop();
            }else{
                continue;
            }
        }

        mini_heap.push(Reverse((val, key)));
    }

    let mut ans = vec![];

    while let Some(cnt) = mini_heap.pop() {
        ans.push(cnt.0.0);
    }
    ans

}